### PR TITLE
Enable to set random ranges to RandomEmojiOverlay parameters

### PR DESCRIPTION
**Pull request by lyakaap**
## Summary
- [x] I have read CONTRIBUTING.md to understand how to contribute to this repository :)

Added randomness to RandomEmojiOverlay parameters like emoji_size.
A motivation is that I wanted to randomly change the parameters on-the-fly, rather than "fixed" like current implementation.
This modified augmentation was actually used in ISC21 Descriptor Track 1st-place solution.

## Test Results

The tests are passed except the following error.

```
Traceback (most recent call last):
  File "/Users/shuhei.yokoo/Documents/AugLy/augly/tests/image_tests/transforms_unit_test.py", line 177, in test_RandomEmojiOverlay
    self.evaluate_class(
  File "/Users/shuhei.yokoo/Documents/AugLy/augly/tests/image_tests/base_unit_test.py", line 129, in evaluate_class
    are_equal_images(dst, ref), "Expected and outputted images do not match"
  File "/Users/shuhei.yokoo/Documents/AugLy/augly/tests/image_tests/base_unit_test.py", line 20, in are_equal_images
    return a.size == b.size and np.allclose(np.array(a), np.array(b))
  File "<__array_function__ internals>", line 5, in allclose
  File "/Users/shuhei.yokoo/.pyenv/versions/augly/lib/python3.8/site-packages/numpy/core/numeric.py", line 2249, in allclose
    res = all(isclose(a, b, rtol=rtol, atol=atol, equal_nan=equal_nan))
  File "<__array_function__ internals>", line 5, in isclose
  File "/Users/shuhei.yokoo/.pyenv/versions/augly/lib/python3.8/site-packages/numpy/core/numeric.py", line 2358, in isclose
    return within_tol(x, y, atol, rtol)
  File "/Users/shuhei.yokoo/.pyenv/versions/augly/lib/python3.8/site-packages/numpy/core/numeric.py", line 2339, in within_tol
    return less_equal(abs(x-y), atol + rtol * abs(y))
ValueError: operands could not be broadcast together with shapes (1080,1920,3) (1080,1920,4) 

----------------------------------------------------------------------
Ran 71 tests in 35.609s

FAILED (errors=1, skipped=4)
sys:1: ResourceWarning: unclosed file <_io.BufferedReader name='/Users/shuhei.yokoo/Documents/AugLy/augly/assets/tests/image/inputs/dfdc_1.jpg'>
```

It seems that expected image has alpha channel whereas output image doesn't have alpha channel.
I'm not sure why it happened. Replacing the expected image is needed?

### CODE REVIEW

File: augly/image/functional.py [hunk 1]
```diff
@@ -1035,12 +1035,13 @@ def overlay_emoji(
...
-    emoji_size: float = 0.15,
-    x_pos: float = 0.4,
-    y_pos: float = 0.8,
+    emoji_size: Union[float, Tuple[float, float]] = 0.15,
+    x_pos: Union[float, Tuple[float, float]] = 0.4,
+    y_pos: Union[float, Tuple[float, float]] = 0.8,
     metadata: Optional[List[Dict[str, Any]]] = None,
     bboxes: Optional[List[Tuple]] = None,
     bbox_format: Optional[str] = None,
+    seed: int = 42,
```
Comment (zpapakipos) [inline augly/image/functional.py:1044]:
```suggestion
    seed: int = 42,
    metadata: Optional[List[Dict[str, Any]]] = None,
    bboxes: Optional[List[Tuple]] = None,
    bbox_format: Optional[str] = None,
```
nit: augmentation-specific args should go before `metadata`, `bboxes`, & `bbox_format`.

Comment (zpapakipos) [inline augly/image/functional.py:1044]:
```suggestion
    seed: int = 42,
    metadata: Optional[List[Dict[str, Any]]] = None,
    bboxes: Optional[List[Tuple]] = None,
    bbox_format: Optional[str] = None,
```
nit: augmentation-specific args should go before metadata, bboxes, & bbox_format.

File: augly/image/functional.py [hunk 2]
```diff
@@ -1073,10 +1080,21 @@ def overlay_emoji(
         specify `bbox_format` if `bboxes` is provided. Supported bbox_format values are
         "pascal_voc", "pascal_voc_norm", "coco", and "yolo"
 
+    @param seed: seed for numpy random generator to select random pixels for shuffling
```
Comment (zpapakipos) [inline augly/image/functional.py:1083]: nit: move this above `metadata`

File: augly/image/functional.py [hunk 3]
```diff
@@ -1073,10 +1080,21 @@ def overlay_emoji(
...
+
     @returns: the augmented PIL Image
     """
     image = imutils.validate_and_load_image(image)
 
+    np.random.seed(seed)
+
+    if isinstance(emoji_size, tuple):
+        emoji_size = np.random.uniform(emoji_size[0], emoji_size[1])
```
Comment (zpapakipos) [inline augly/image/functional.py:1092]:
```suggestion
        assert emoji_size[0] < emoji_size[1], (
            "emoji_size must be a float or a tuple [low, high) to sample the value from"
        )
        emoji_size = np.random.uniform(emoji_size[0], emoji_size[1])
```
Give a useful error message if a bad tuple is given

File: augly/image/functional.py [hunk 4]
```diff
@@ -1073,10 +1080,21 @@ def overlay_emoji(
...
+    if isinstance(x_pos, tuple):
+        x_pos = np.random.uniform(x_pos[0], x_pos[1])
```
Comment (zpapakipos) [inline augly/image/functional.py:1094]:
```suggestion
        assert x_pos[0] < x_pos[1], (
            "x_pos must be a float or a tuple [low, high) to sample the value from"
        )
        x_pos = np.random.uniform(x_pos[0], x_pos[1])
```
+1 from above

File: augly/image/functional.py [hunk 5]
```diff
@@ -1073,10 +1080,21 @@ def overlay_emoji(
...
+    if isinstance(y_pos, tuple):
+        y_pos = np.random.uniform(y_pos[0], y_pos[1])
```
Comment (zpapakipos) [inline augly/image/functional.py:1096]:
```suggestion
        assert y_pos[0] < y_pos[1], (
            "y_pos must be a float or a tuple [low, high) to sample the value from"
        )
        y_pos = np.random.uniform(y_pos[0], y_pos[1])
```
+1 from above

File: augly/image/transforms.py [hunk 6]
```diff
@@ -2340,6 +2351,7 @@ def apply_transform(
             metadata=metadata,
             bboxes=bboxes,
             bbox_format=bbox_format,
+            seed=self.seed,
```
Comment (zpapakipos) [inline augly/image/transforms.py:2354]:
```suggestion
            seed=self.seed,
            metadata=metadata,
            bboxes=bboxes,
            bbox_format=bbox_format,
```
tiny nit

File: augly/image/functional.py [hunk 7]
```diff
@@ -1055,11 +1056,17 @@ def overlay_emoji(
 
     @param opacity: the lower the opacity, the more transparent the overlaid emoji
 
-    @param emoji_size: size of the emoji is emoji_size * height of the original image
+    @param emoji_size: size of the emoji is emoji_size * height of the original image.
+        If set to tuple, value will be randomly chosen from the range of the first value
+        to the second value
```
Comment (zpapakipos) [inline augly/image/functional.py:1061]:
```suggestion
        If set to a tuple, the value will be randomly chosen from that range [low, high)
```

File: augly/image/functional.py [hunk 8]
```diff
@@ -1055,11 +1056,17 @@ def overlay_emoji(
...
 
-    @param x_pos: position of emoji relative to the image width
+    @param x_pos: position of emoji relative to the image width.
+        If set to tuple, value will be randomly chosen from the range of the first value
+        to the second value
```
Comment (zpapakipos) [inline augly/image/functional.py:1065]:
```suggestion
    @param x_pos: position of emoji relative to the image width. If set to a tuple, the
        value will be randomly chosen from that range [low, high)
```

File: augly/image/functional.py [hunk 9]
```diff
@@ -1055,11 +1056,17 @@ def overlay_emoji(
...
 
-    @param y_pos: position of emoji relative to the image height
+    @param y_pos: position of emoji relative to the image height.
+        If set to tuple, value will be randomly chosen from the range of the first value
+        to the second value
```
Comment (zpapakipos) [inline augly/image/functional.py:1069]:
```suggestion
    @param y_pos: position of emoji relative to the image height. If set to tuple, value
        will be randomly chosen from that range [low, high)
```

**Review by zpapakipos — CHANGES_REQUESTED** [review]
Hi @lyakaap, thanks so much for this PR! :)

I left a couple comments with small changes I'd like before landing this. You **will** indeed need to replace the expected output file for the `RandomEmojiOverlay` unit test. You can do this e.g. by temporarily adding a line `pil_dst.save("augly/assets/tests/image/dfdc_expected_output/test_RandomEmojiOverlay.png")` [here](https://github.com/facebookresearch/AugLy/blob/e287b4e5abc994ac0b52723c908b65ac0f4219c9/augly/tests/image_tests/base_unit_test.py#L94) and then running `python -m unittest augly.tests.image_tests.functional_unit_test.FunctionalImageUnitTest.test_RandomEmojiOverlay`. Then if that passes, remove the line you added in `base_unit_test.py` and commit your changes (should just be the new `test_RandomEmojiOverlay.png` file)!

**Comment by lyakaap** [comment]
Thank you for your quick & kind review!

I reflected your review and confirmed the tests are passed.
Like you instructed, I rewrote png file by inserting the line, but in `evaluate_class` instead `evaluate_function`.
And executed `python -m unittest augly.tests.image_tests.transforms_unit_test.TransformsImageUnitTest.test_RandomEmojiOverlay`.

**Review by zpapakipos — CHANGES_REQUESTED** [review]
Sorry for the delay with reviewing this PR, I was on PTO for the holidays! :) Please fix the nits I suggested below, and also run `black` on the files which are failing in the lint workflow, as you can see.
```
black augly/image/functional.py
black augly/tests/image_tests/transforms_unit_test.py
```

**Comment by lyakaap** [comment]
No problem :)
I pushed your suggested changes. Please check.

**Comment by zpapakipos** [comment]
Hi @lyakaap, sorry for the delay! I was pretty sure your changes did not cause the test failures you're seeing in the `test_python` workflow here, so I went ahead & debugged them, here's the fix: https://github.com/facebookresearch/AugLy/pull/185. I'll land it ASAP, then you can rebase onto main & we'll ensure all your tests pass here before landing :)

**Comment by lyakaap** [comment]
Thanks @zpapakipos.
I rebased the PR and force pushed here. Please let me know if there are any problems.

File: augly/image/transforms.py [hunk 10]
```diff
@@ -2318,6 +2318,7 @@ def apply_random_transform(
...
```
Comment (zpapakipos) [inline augly/image/transforms.py:2321]:
```suggestion
```

File: augly/image/transforms.py [hunk 11]
```diff
@@ -2326,21 +2326,31 @@ def __init__(
...
     ):
         """
         @param emoji_directory: iopath directory uri containing the emoji images
 
         @param opacity: the lower the opacity, the more transparent the overlaid emoji
 
-        @param emoji_size: size of the emoji is emoji_size * height of the original image
+        @param emoji_size: size of the emoji is emoji_size * height of the original image.
+            If set to tuple, value will be randomly chosen from the range of the first value
+            to the second value
```
Comment (jbitton) [inline augly/image/transforms.py:2342]:
```suggestion
             If set to a tuple, a size will randomly be chosen from the range provided
```

File: augly/image/transforms.py [hunk 12]
```diff
@@ -2326,21 +2326,31 @@ def __init__(
...
 
-        @param x_pos: position of emoji relative to the image width
+        @param x_pos: position of emoji relative to the image width.
+            If set to tuple, value will be randomly chosen from the range of the first value
+            to the second value
```
Comment (jbitton) [inline augly/image/transforms.py:2346]:
```suggestion
            If set to a tuple, a position will randomly be chosen from the range provided
```

File: augly/image/transforms.py [hunk 13]
```diff
@@ -2326,21 +2326,31 @@ def __init__(
...
 
-        @param y_pos: position of emoji relative to the image height
+        @param y_pos: position of emoji relative to the image height.
+            If set to tuple, value will be randomly chosen from the range of the first value
+            to the second value
```
Comment (jbitton) [inline augly/image/transforms.py:2350]:
```suggestion
            If set to a tuple, a position will randomly be chosen from the range provided
```

File: augly/image/transforms.py [hunk 14]
```diff
@@ -2386,6 +2397,7 @@ def apply_transform(
             emoji_size=self.emoji_size,
             x_pos=self.x_pos,
             y_pos=self.y_pos,
+            seed=self.seed,
```
Comment (jbitton) [inline augly/image/transforms.py:2400]: shouldn't you set the random seed on line 2392, i.e. where we randomly choose which emoji to use?

File: augly/image/functional.py [hunk 15]
```diff
@@ -1056,11 +1057,17 @@ def overlay_emoji(
...
+    @param x_pos: position of emoji relative to the image width. If set to a tuple, the
+        value will be randomly chosen from that range [low, high)
 
-    @param x_pos: position of emoji relative to the image width
+    @param y_pos: position of emoji relative to the image height. If set to tuple, value
+        will be randomly chosen from that range [low, high)
 
-    @param y_pos: position of emoji relative to the image height
+    @param seed: if provided, this will set the random seed to ensure
+            consistency between runs
```
Comment (jbitton) [inline augly/image/functional.py:1070]: if we decide to continue to set the random parameters here instead of transforms.py, please ensure all docstrings are identical across both files. otherwise, revert these changes!

File: augly/image/functional.py [hunk 16]
```diff
@@ -1036,9 +1036,10 @@ def overlay_emoji(
     output_path: Optional[str] = None,
     emoji_path: str = utils.EMOJI_PATH,
     opacity: float = 1.0,
-    emoji_size: float = 0.15,
-    x_pos: float = 0.4,
-    y_pos: float = 0.8,
+    emoji_size: Union[float, Tuple[float, float]] = 0.15,
```
Comment (jbitton) [inline augly/image/functional.py:1039]: overall question: so currently, `overlay_emoji` is *not* a random function, however `RandomEmojiOverlay` is. `RandomOverlayEmoji` currently randomly chooses an emoji to overlay, however the code here lets `overlay_emoji` do the randomness.  thus, the value setting is spread across two functions -- we want to avoid this.

personally, my suggestion would be to revert these changes in `functional.py` and allow the additional random parameter/seed setting to occur in `transforms.py` :) thoughts? 

cc: @zpapakipos

**Review by jbitton — CHANGES_REQUESTED** [review]
Thank you so much for your PR and apologies for the delayed review! I have a couple overall comments about your changes, let me know what you think!

File: augly/image/functional.py (hunk 16 above)
Comment (zpapakipos) [reply augly/image/functional.py:1039]: Yes, I think I agree that would be the best. I thought about this before too, and I didn't suggest that change just because I feel like ideally at some point we should add random sampling abilities for all the transform classes -- e.g. be able to pass in ranges for each param, or have a method `sample_parameters()` which can be called after `__init__()` before `__call__()` with a seed.

For now, indeed I think it's better in this PR to take in the param ranges in `RandomOverlayEmoji` but only take in the params as floats in `overlay_emoji()`. So all the changes in `functional.py` should be reverted, and the seed setting & param sampling should be moved to right after line 2392 in `transforms.py`.

File: augly/image/transforms.py (hunk 14 above)
Comment (zpapakipos) [reply augly/image/transforms.py:2400]: Agreed, @lyakaap when you move all the param setting to line 2392, please make sure to set the seed before line 2392 so that *all* the param sampling is reproducible using the seed.

**Comment by zpapakipos** [comment]
Hi @lyakaap, did you have a chance to look at @jbitton's requested changes on this PR? If you don't have time/bandwidth to make them please let me know, but otherwise if you can make those changes I would appreciate it so we can land the PR!

**Comment by lyakaap** [comment]
Sorry for the delay, and thank you for the review!
I reflected your review on this PR. Please check again if it works.

File: augly/utils/expected_output/image_tests/expected_metadata.json [hunk 17]
```diff
@@ -152,7 +152,8 @@
             "src_height": 1080,
             "src_width": 1920,
             "x_pos": 0.4,
-            "y_pos": 0.8
+            "y_pos": 0.8,
+            "seed": 42
```
Comment (zpapakipos) [inline augly/utils/expected_output/image_tests/expected_metadata.json:156]:
```suggestion
            "y_pos": 0.8
```

File: augly/utils/expected_output/image_tests/expected_metadata.json [hunk 18]
```diff
@@ -318,7 +319,8 @@
...
```
Comment (zpapakipos) [inline augly/utils/expected_output/image_tests/expected_metadata.json:323]:
```suggestion
            "y_pos": 0.8
```

File: augly/image/transforms.py [hunk 19]
```diff
@@ -2378,14 +2386,39 @@ def apply_transform(
...
+            x_pos = random.uniform(self.x_pos[0], self.x_pos[1])
+        else:
+            x_pos = self.x_pos
+        if isinstance(self.y_pos, tuple):
+            assert (
+                self.y_pos[0] < self.y_pos[1]
+            ), "y_pos must be a float or a tuple [low, high) to sample the value from"
+            y_pos = random.uniform(self.y_pos[0], self.y_pos[1])
+        else:
+            y_pos = self.y_pos
```
Comment (zpapakipos) [inline augly/image/transforms.py:2412]:
```suggestion
        assert isinstance(self.emoji_size, (float, int)) or (
            isinstance(self.emoji_size, tuple) and self.emoji_size[0] < self.emoji_size[1]
        ), "emoji_size must be a float or a tuple [low, high) to sample the value from"
        assert isinstance(self.x_pos, (float, int)) or (
            isinstance(self.x_pos, tuple) and self.x_pos[0] < self.x_pos[1]
        ), "x_pos must be a float or a tuple [low, high) to sample the value from"
        assert isinstance(self.y_pos, (float, int)) or (
            isinstance(self.y_pos, tuple) and self.y_pos[0] < self.y_pos[1]
        ), "y_pos must be a float or a tuple [low, high) to sample the value from"
        
        emoji_size = (
            random.uniform(self.emoji_size[0], self.emoji_size[1])
            if isinstance(self.emoji_size, tuple)
            else self.emoji_size
        )
        x_pos = (
            random.uniform(self.x_pos[0], self.x_pos[1])
            if isinstance(self.x_pos, tuple)
            else self.x_pos
        )
        y_pos = (
            random.uniform(self.y_pos[0], self.y_pos[1])
            if isinstance(self.y_pos, tuple)
            else self.y_pos
        )
```

File: augly/image/transforms.py [hunk 20]
```diff
@@ -2378,14 +2386,43 @@ def apply_transform(
 
         @returns: Augmented PIL Image
         """
+        if self.seed is not None:
+            random.seed(self.seed)
+
+        assert isinstance(self.emoji_size, (float, int)) or (
+            isinstance(self.emoji_size, tuple) and self.emoji_size[0] < self.emoji_size[1]
```
Comment (zpapakipos) [inline augly/image/transforms.py:2393]:
```suggestion
            isinstance(self.emoji_size, tuple)
            and self.emoji_size[0] < self.emoji_size[1]
```

File: augly/image/transforms.py [hunk 21]
```diff
@@ -2326,21 +2326,28 @@ def __init__(
...
         p: float = 1.0,
     ):
         """
         @param emoji_directory: iopath directory uri containing the emoji images
 
         @param opacity: the lower the opacity, the more transparent the overlaid emoji
 
-        @param emoji_size: size of the emoji is emoji_size * height of the original image
+        @param emoji_size: size of the emoji is emoji_size * height of the original image.
+            If set to a tuple, a position will randomly be chosen from the range provided
```
Comment (zpapakipos) [inline augly/image/transforms.py:2341]:
```suggestion
        @param emoji_size: size of the emoji is emoji_size * height of the original
            image. If set to a tuple, a position will randomly be chosen from the range
            provided
```

File: augly/image/transforms.py [hunk 22]
```diff
@@ -2326,21 +2326,28 @@ def __init__(
...
 
-        @param x_pos: position of emoji relative to the image width
+        @param x_pos: position of emoji relative to the image width.
+            If set to a tuple, a position will randomly be chosen from the range provided
```
Comment (zpapakipos) [inline augly/image/transforms.py:2344]:
```suggestion
        @param x_pos: position of emoji relative to the image width. If set to a tuple, a
            position will randomly be chosen from the range provided
```

File: augly/image/transforms.py [hunk 23]
```diff
@@ -2326,21 +2326,28 @@ def __init__(
...
 
-        @param y_pos: position of emoji relative to the image height
+        @param y_pos: position of emoji relative to the image height.
+            If set to a tuple, a position will randomly be chosen from the range provided
```
Comment (zpapakipos) [inline augly/image/transforms.py:2347]:
```suggestion
        @param y_pos: position of emoji relative to the image height. If set to a tuple, a
            position will randomly be chosen from the range provided
```

File: augly/image/transforms.py [hunk 24]
```diff
@@ -2326,21 +2326,28 @@ def __init__(
...
+
+        @param seed: if provided, this will set the random seed to ensure
+            consistency between runs
```
Comment (zpapakipos) [inline augly/image/transforms.py:2350]:
```suggestion
        @param seed: if provided, this will set the random seed to ensure consistency
            between runs
```

File: augly/image/transforms.py [hunk 25]
```diff
@@ -2378,14 +2386,44 @@ def apply_transform(
...
+        assert isinstance(self.emoji_size, (float, int)) or (
+            isinstance(self.emoji_size, tuple)
+            and self.emoji_size[0] < self.emoji_size[1]
+        ), "emoji_size must be a float or a tuple [low, high) to sample the value from"
+        assert isinstance(self.x_pos, (float, int)) or (
+            isinstance(self.x_pos, tuple) and self.x_pos[0] < self.x_pos[1]
+        ), "x_pos must be a float or a tuple [low, high) to sample the value from"
+        assert isinstance(self.y_pos, (float, int)) or (
+            isinstance(self.y_pos, tuple) and self.y_pos[0] < self.y_pos[1]
+        ), "y_pos must be a float or a tuple [low, high) to sample the value from"
```
Comment (zpapakipos) [inline augly/image/transforms.py:2401]:
```suggestion
        assert isinstance(self.emoji_size, (float, int)) or (
            isinstance(self.emoji_size, tuple)
            and self.emoji_size[0] < self.emoji_size[1]
        ), "emoji_size must be a float or a tuple [low, high) to sample the value from"
        assert isinstance(self.x_pos, (float, int)) or (
            isinstance(self.x_pos, tuple) and self.x_pos[0] < self.x_pos[1]
        ), "x_pos must be a float or a tuple [low, high) to sample the value from"
        assert isinstance(self.y_pos, (float, int)) or (
            isinstance(self.y_pos, tuple) and self.y_pos[0] < self.y_pos[1]
        ), "y_pos must be a float or a tuple [low, high) to sample the value from"

        if self.seed is not None:
            random.seed(self.seed)
```
